Support ADC in Lucene-on-faiss initial commit

finnroblin opened the pull request:
### Description
Draft implementation to integrate asymmetric distance calculation (ADC) with lucene-on-faiss.

This implementation uses the following low-level approach:
- Within the `NativeEngines990KnnVectorsReader` see from the `FieldInfo` if adc is enabled or not. We need to check here so we can obtain the original `SpaceType`, as this information is lost for binary indices when it is changed to hamming.
- Add new split within `trySearchWithMemoryOptimizedSearch` on if adc is enabled or not.
- Add `isAdc` parameter to `createVectorSearcher` method.
- Add new method `searchWithAdc`. The reason for a new method is that preexisting scorer supplier logic depends on the query and document vector types matching (`float/float` or `byte/byte`), whereas ADC is `float/byte`. 
- Create new ADC scorer, leveraging java logic originally implemented for ADC exact search.

### Related Issues
Resolves #2714 

### Check List
- [X] New functionality includes testing.
- [x] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

finnroblin commented:
Will add changelog entry to next revision.

0ctopus13prime commented on src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsReader.java:
```diff
@@ -364,13 +414,31 @@ private IOSupplier<VectorSearcher> getVectorSearcherSupplier(final FieldInfo fie
 
+    private AdcConfig getAdcConfig(String field) {
+        boolean isAdcEnabled = adcEnabledCache.computeIfAbsent(field, f -> {
+            FieldInfo fieldInfo = segmentReadState.fieldInfos.fieldInfo(f);
```
Do we need to wrap this with try-catch and return false whenever an exception to be thrown?

finnroblin commented:
Thanks @shatejas for the review. I understand the concern around breaking the FlatVectorScorer interface with the `searchWithAdc` method. 

I implemented a different approach in my latest commit that instantiates a `ADCFlatVectorsScorer` through the `FaissMemoryOptimizedSearcherFactory.createVectorSearcher` method. The factory is passed `FieldInfo` and then possibly creates an ADC scorer based on the quantization state. I believe that this is relatively efficient... since the vector searcher already makes a costly `Directory.openInput` call, it is not too much extra overhead to also grab adc info from quantization state. My hope is this change also addresses your comments about caching.

finnroblin commented on src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsReader.java:
```diff
@@ -63,11 +68,23 @@ public class NativeEngines990KnnVectorsReader extends KnnVectorsReader {
     private volatile Map<String, VectorSearcherHolder> vectorSearchers;
+    private final Map<String, Boolean> adcEnabledCache;
+
+    private static class AdcConfig {
```
Moved to searcher.

finnroblin commented on src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissMemoryOptimizedSearcher.java:
```diff
@@ -34,18 +35,27 @@ public class FaissMemoryOptimizedSearcher implements VectorSearcher {
         }
+        // spaceType is only used if ADC is enabled.
+        this.isAdc = isAdc;
+        this.spaceType = spaceType;
```
No, will remove spaceType.

shatejas commented on src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FlatVectorsScorerProvider.java:
```diff
@@ -14,22 +14,131 @@
+            }
+
+            return switch (spaceType) {
+                case HAMMING -> throw new IllegalArgumentException("hamming distance unsupported by ADC");
```
nit: probably not needed, let it go to default and change the message in default to "spaceType not supported for ADC"

shatejas reviewed (commented):
Thanks for all the changes, this is looking good. waiting for the CIs to pass before approving
